src/librbd/mirror/snapshot/GetImageStateRequest.cc

```diff
@@ -0,0 +1,114 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/mirror/snapshot/GetImageStateRequest.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/Utils.h"
+#include "librbd/mirror/snapshot/Types.h"
+#include "librbd/mirror/snapshot/Utils.h"
+
+#define dout_subsys ceph_subsys_rbd
+
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::mirror::snapshot::GetImageStateRequest: " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace mirror {
+namespace snapshot {
+
+using librbd::util::create_rados_callback;
+
+template <typename I>
+void GetImageStateRequest<I>::send() {
+  read_object();
+}
+
+
+template <typename I>
+void GetImageStateRequest<I>::read_object() {
+  CephContext *cct = m_image_ctx->cct;
+
+  auto oid = util::image_state_object_name(m_image_ctx, m_snap_id,
+                                           m_object_index);
+  ldout(cct, 20) << oid << dendl;
+
+  librados::ObjectReadOperation op;
+  m_bl.clear();
+  op.read(0, 0, &m_bl, nullptr);
+
+  librados::AioCompletion *comp = create_rados_callback<
+    GetImageStateRequest<I>,
+    &GetImageStateRequest<I>::handle_read_object>(this);
+  int r = m_image_ctx->md_ctx.aio_operate(oid, comp, &op, nullptr);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void GetImageStateRequest<I>::handle_read_object(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to read image state object: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  if (m_object_index == 0) {
+    ImageStateHeader header;
+    bufferlist bl;
+    bl.substr_of(m_bl, 0, header.length());
+    try {
+      using ceph::decode;
+      decode(header, bl);
+    } catch (const buffer::error &err) {
+      lderr(cct) << "failed to decode image state object header" << dendl;
+      finish(-EBADMSG);
+      return;
+    }
+    m_object_count = header.object_count;
+    m_state_bl.substr_of(m_bl, header.length(),
```

Comment:
... and then you would just append from the new iterator position until the end of the bufferlist.